### PR TITLE
fixes #23454; IndexDefect thrown when destructuring a lent tuple

### LINKED ISSUE #23454: IndexDefect thrown when destructuring a lent value
### Description

The following code throws a compiler error, when it is expected to work:

```nim
type
    Letter = enum A

    LetterPairs = object
        values: seq[(Letter, string)]

iterator items(list: var LetterPairs): lent (Letter, string) =
    for item in list.values:
        yield item

var instance = LetterPairs(values: @[ (A, "foo") ])

for (a, _) in instance:
    case a
    of A: discard
```

### Nim Version

```
Nim Compiler Version 2.0.2 [Linux: amd64]
Compiled at 2023-12-15
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: c4c44d10df8a14204a75c34e499def200589cb7c
active boot switches: -d:release
```

And also:

```
> ~/Code/Nim/compiler/nim --version       
Nim Compiler Version 2.1.1 [Linux: amd64]
Compiled at 2024-03-14
Copyright (c) 2006-2024 by Andreas Rumpf

git hash: 7657a637b8500e7ab95c18c8559d66cc2da2d124
active boot switches:
```

### Current Output

```text
The thrown error is:


Hint: used config file '/home/nycto/Code/Nim/config/nim.cfg' [Conf]
Hint: used config file '/home/nycto/Code/Nim/config/config.nims' [Conf]
...................................................................../home/nycto/Code/Nim/compiler/nim.nim(169) nim
/home/nycto/Code/Nim/compiler/nim.nim(124) handleCmdLine
/home/nycto/Code/Nim/compiler/main.nim(338) mainCommand
/home/nycto/Code/Nim/compiler/main.nim(304) compileToBackend
/home/nycto/Code/Nim/compiler/main.nim(142) commandCompileToC
/home/nycto/Code/Nim/compiler/pipelines.nim(325) compilePipelineProject
/home/nycto/Code/Nim/compiler/pipelines.nim(245) compilePipelineModule
/home/nycto/Code/Nim/compiler/pipelines.nim(188) processPipelineModule
/home/nycto/Code/Nim/compiler/pipelines.nim(26) processPipeline
/home/nycto/Code/Nim/compiler/cgen.nim(2123) genTopLevelStmt
/home/nycto/Code/Nim/compiler/transf.nim(1244) transformStmt
/home/nycto/Code/Nim/compiler/transf.nim(1160) processTransf
/home/nycto/Code/Nim/compiler/transf.nim(1139) transform
/home/nycto/Code/Nim/compiler/transf.nim(109) transformSons
/home/nycto/Code/Nim/compiler/transf.nim(1031) transform
/home/nycto/Code/Nim/compiler/transf.nim(804) transformFor
/home/nycto/Code/Nim/compiler/transf.nim(1038) transform
/home/nycto/Code/Nim/compiler/transf.nim(269) transformBlock
/home/nycto/Code/Nim/compiler/transf.nim(109) transformSons
/home/nycto/Code/Nim/compiler/transf.nim(1139) transform
/home/nycto/Code/Nim/compiler/transf.nim(109) transformSons
/home/nycto/Code/Nim/compiler/transf.nim(1139) transform
/home/nycto/Code/Nim/compiler/transf.nim(109) transformSons
/home/nycto/Code/Nim/compiler/transf.nim(1038) transform
/home/nycto/Code/Nim/compiler/transf.nim(269) transformBlock
/home/nycto/Code/Nim/compiler/transf.nim(109) transformSons
/home/nycto/Code/Nim/compiler/transf.nim(1036) transform
/home/nycto/Code/Nim/compiler/transf.nim(293) transformWhile
/home/nycto/Code/Nim/compiler/transf.nim(109) transformSons
/home/nycto/Code/Nim/compiler/transf.nim(1139) transform
/home/nycto/Code/Nim/compiler/transf.nim(109) transformSons
/home/nycto/Code/Nim/compiler/transf.nim(1139) transform
/home/nycto/Code/Nim/compiler/transf.nim(109) transformSons
/home/nycto/Code/Nim/compiler/transf.nim(1105) transform
/home/nycto/Code/Nim/compiler/transf.nim(448) transformYield
/home/nycto/Code/Nim/compiler/lowerings.nim(34) newTupleAccess
/home/nycto/Code/Nim/lib/system/fatal.nim(53) sysFatal
Error: unhandled exception: index 1 not in 0 .. 0 [IndexDefect]
```
```


### Expected Output

```text
I expect this code to compile, or at the very least produce an actionable error message
```

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ringabout**
fixes #23454

### CODE REVIEW

**Comment by Araq** [comment]
Adhoc solution plus it's unclear to me why `var T` would not have the same problem.

**Comment by ringabout** [comment]
Well, `var T` does have the similar problems

without `mitems`
```nim
block: # bug #23454
  type
    Letter = enum
      A

    LetterPairs = object
      values: seq[(Letter, string)]

  iterator items(list: var LetterPairs): var (Letter, string) =
    for item in list.values:
      yield item

  var instance = LetterPairs(values: @[(A, "foo")])

  for (a, _) in instance:
    case a
    of A: discard
```

gives `expression has no address`, I suppose there is a hidden address for `var/lent`

```nim
block: # bug #23454
  type
    Letter = enum
      A

    LetterPairs = object
      values: seq[(Letter, string)]

  iterator items(list: var LetterPairs): var (Letter, string) =
    for item in list.values.mitems:
      yield item

  var instance = LetterPairs(values: @[(A, "foo")])

  for (a, _) in instance:
    case a
    of A: discard
```

```
Error: unhandled exception: index 1 not in 0 .. 0 [IndexDefect]
```

File: compiler/transf.nim [hunk 1]
```diff
@@ -441,11 +441,20 @@ proc transformYield(c: PTransf, n: PNode): PNode =
...
+        if e.typ.kind == tyPtr: # tyLent
+          tmp = newTreeIT(nkHiddenDeref, e.info, e.typ.skipTypes({tyPtr}),
+                  tmp)
+
         result.add transform(c, v)
         for i in 0..<c.transCon.forStmt[0].len-1:
           let lhs = c.transCon.forStmt[0][i]
           let rhs = transform(c, newTupleAccess(c.graph, tmp, i))
-          result.add(asgnTo(lhs, rhs))
+          if lhs.typ.kind == tyLent:
```
Comment (ringabout) [inline compiler/transf.nim]:
```suggestion
          if lhs.typ.kind in {tyLent, tyVar}:
```

File: compiler/transf.nim [hunk 2]
```diff
@@ -441,11 +441,20 @@ proc transformYield(c: PTransf, n: PNode): PNode =
         let v = newNodeI(nkVarSection, e.info)
         v.addVar(tmp, e)
 
+        if e.typ.kind == tyPtr: # tyLent
```
Comment (ringabout) [inline compiler/transf.nim]:
```suggestion
        if e.typ.kind == tyPtr: # tyLent, tyVar
```

**Comment by ringabout** [comment]
Succeeded by https://github.com/nim-lang/Nim/pull/23963

**Pull request closed without merging**